src/core/shards/logging.cpp

```diff
@@ -100,30 +103,162 @@ struct Msg : public LoggingBase {
+      _formatted_cache.clear();
+      _formatted_cache.reserve(n_items);
+      for (size_t i = 0; i < n_items; i++) {
+        spdlog::memory_buf_t formatted;
+        spdlog::sinks::base_sink<Mutex>::formatter_->format(q_.at(i), formatted);
+        _formatted_cache.push_back(fmt::to_string(formatted));
+      }
+    }
+
+    return _formatted_cache;
+  }
+
+protected:
+  void sink_it_(const spdlog::details::log_msg &msg) override {
+    if (msg.level >= min_level_) {
+      _dirty = true;
+      q_.push_back(spdlog::details::log_msg_buffer{msg});
+    }
+  }
+  void flush_() override {}
+
+private:
+  spdlog::details::circular_q<spdlog::details::log_msg_buffer> q_;
+  spdlog::level::level_enum min_level_;
+
+  bool _dirty;
+  std::vector<std::string> _formatted_cache;
+};
+
+using custom_ringbuffer_sink_mt = custom_ringbuffer_sink<std::mutex>;
+using custom_ringbuffer_sink_st = custom_ringbuffer_sink<spdlog::details::null_mutex>;
+
+struct CaptureLog {
+  static SHTypesInfo inputTypes() { return CoreInfo::NoneType; }
+
+  static SHTypesInfo outputTypes() { return CoreInfo::StringSeqType; }
+
+  static SHParametersInfo parameters() { return _params; }
+
+  void setParam(int index, const SHVar &value) {
+    switch (index) {
+    case 0:
+      _n_items = size_t(value.payload.intValue);
+      break;
+    case 1:
+      _min_level = value.payload.stringValue;
+      break;
+    case 2:
+      _pattern = value.payload.stringValue;
+      break;
+    case 3:
+      _suspend = value.payload.stringValue;
+      break;
+    default:
+      break;
+    }
+  }
+
+  SHVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return Var(int64_t(_n_items));
+    case 1:
+      return Var(_min_level);
+    case 2:
+      return Var(_pattern);
+    case 3:
+      return Var(_suspend);
+    default:
+      return Var::Empty;
+    }
+  }
+
+  void warmup(SHContext *context) {
+    auto logger = spdlog::default_logger();
+    assert(logger);
+    auto sink = logger->sinks().at(0);
+    assert(sink);
+    auto dist_sink = std::dynamic_pointer_cast<spdlog::sinks::dist_sink_mt>(sink);
+    if (dist_sink) {
+      _ring = std::make_shared<custom_ringbuffer_sink_mt>(_n_items, spdlog::level::from_str(_min_level));
+      _ring->set_formatter(
+          std::unique_ptr<spdlog::formatter>(new spdlog::pattern_formatter(_pattern, spdlog::pattern_time_type::local)));
+      dist_sink->add_sink(_ring);
+    }
+  }
+
+  void cleanup() {
+    auto logger = spdlog::default_logger();
+    assert(logger);
+    auto sink = logger->sinks().at(0);
+    assert(sink);
+    auto dist_sink = std::dynamic_pointer_cast<spdlog::sinks::dist_sink_mt>(sink);
+    if (dist_sink) {
+      dist_sink->remove_sink(_ring);
+      _ring.reset();
+    }
+  }
+
+  SHVar activate(SHContext *context, const SHVar &input) {
+    if (likely((bool)_ring)) {
+      while (_suspend && !_ring->is_dirty()) {
+        SH_SUSPEND(context, 0);
+      }
+
+      if (_ring->is_dirty()) {
```

Comment:
I think we can avoid checking dirty flag twice here (as it potentially is atomic it will introduce barriers so better do once)

Reply:
Good point.

Reply:
Actually it is needed to do so when  `_suspend` is `false`.

However, since the `:Suspend` parameter doesn't change after it is assigned, maybe it should use two different `activate` overrides.

Reply:
yeah